rust/windows-client/src/settings.ts

```diff
@@ -0,0 +1,225 @@
+
+// Lock the UI when we're saving to disk, since disk writes are technically async.
+function lockAdvancedSettingsForm() {
+  authBaseUrlInput.disabled = true;
+  apiUrlInput.disabled = true;
+  logFilterInput.disabled = true;
+  resetAdvancedSettingsBtn.disabled = true;
+  applyAdvancedSettingsBtn.disabled = true;
+
+  resetAdvancedSettingsBtn.textContent = "Updating...";
+  applyAdvancedSettingsBtn.textContent = "Updating...";
+}
+
+function unlockAdvancedSettingsForm() {
+  authBaseUrlInput.disabled = false;
+  apiUrlInput.disabled = false;
+  logFilterInput.disabled = false;
+  resetAdvancedSettingsBtn.disabled = false;
+  applyAdvancedSettingsBtn.disabled = false;
+
+  resetAdvancedSettingsBtn.textContent = "Reset to Defaults";
+  applyAdvancedSettingsBtn.textContent = "Apply";
+}
+
+function lockLogsForm() {
+  exportLogsBtn.disabled = true;
+  clearLogsBtn.disabled = true;
+}
+
+function unlockLogsForm() {
+  exportLogsBtn.disabled = false;
+  clearLogsBtn.disabled = false;
+}
+
+async function applyAdvancedSettings() {
+  console.log("Applying advanced settings");
+  lockAdvancedSettingsForm();
+
+  invoke("apply_advanced_settings", {
+    settings: {
+      auth_base_url: authBaseUrlInput.value,
+      api_url: apiUrlInput.value,
+      log_filter: logFilterInput.value,
+    },
+  })
+    .catch((e: Error) => {
+      console.error(e);
+    })
+    .finally(() => {
+      unlockAdvancedSettingsForm();
+    });
+}
+
+async function resetAdvancedSettings() {
+  console.log("Resetting advanced settings");
+  lockAdvancedSettingsForm();
+
+  invoke("reset_advanced_settings")
+    .then((settings: Settings) => {
+      authBaseUrlInput.value = settings.auth_base_url;
+      apiUrlInput.value = settings.api_url;
+      logFilterInput.value = settings.log_filter;
+    })
+    .catch((e: Error) => {
+      console.error(e);
+    })
+    .finally(() => {
+      unlockAdvancedSettingsForm();
+    });
+}
+
+async function getAdvancedSettings() {
+  console.log("Getting advanced settings");
+  lockAdvancedSettingsForm();
+
+  invoke("get_advanced_settings")
+    .then((settings: Settings) => {
+      authBaseUrlInput.value = settings.auth_base_url;
+      apiUrlInput.value = settings.api_url;
+      logFilterInput.value = settings.log_filter;
+    })
+    .catch((e: Error) => {
+      console.error(e);
+    })
+    .finally(() => {
+      unlockAdvancedSettingsForm();
+    });
+}
+
+async function exportLogs() {
+  console.log("Exporting logs");
+  lockLogsForm();
+
+  invoke("export_logs")
+    .catch((e: Error) => {
+      console.error(e);
+    })
+    .finally(() => {
+      unlockLogsForm();
+    });
+}
+
+async function clearLogs() {
+  console.log("Clearing logs");
+  lockLogsForm();
+
+  invoke("clear_logs")
+    .catch((e: Error) => {
+      console.error(e);
+    })
+    .finally(() => {
+      logCountOutput.innerText = "0 files, 0 MB";
+      unlockLogsForm();
+    });
+}
+
+async function countLogs() {
+  invoke("count_logs")
```

Comment:
So since this is async, JS can run other tasks concurrently, right?
But there's no way to cancel it?

Reply:
Typically yeah. Depends how slow the disk is too, whether it's under load or failing or something. I just don't like filesystems that much